File: client/lib/services/article_service.dart
import 'dart:io';

import 'package:client/datas/models/article.dart';
import 'package:client/datas/responses/article_response.dart';
import 'package:client/datas/responses/articles_response.dart';
import 'package:client/datas/responses/boolean_response.dart';
import 'package:client/datas/responses/id_response.dart';
import 'package:client/http/http.dart';

final _http = Http();

class ArticleService {
  Future<List<Article>> getArticles() async {
    var res = await _http.get("article");
    ArticlesResponse articlesResponse = ArticlesResponse.fromJson(res.response);
    return articlesResponse.articles;
  }

  Future<Article> getArticle(int articleId) async {
    var res = await _http.get("article/" + articleId.toString());
    ArticleResponse articleResponse = ArticleResponse.fromJson(res.response);
    return articleResponse.articleDetails;
  }

  Future<bool> isUniqueTitle(String title) async {
    var res = await _http.post("article/isUniqueTitle", data: {"Title": title});
    BooleanResponse articleResponse = BooleanResponse.fromJson(res.response);
    return articleResponse.value;
  }

  Future<int> createArticle(String title, String description, String content,
      List<int> tagIds) async {
    var res = await _http.post("article/createArticle",
        data: {
          "title": title,
          "description": description,
          "content": content,
          "tagIds": tagIds
        },
        queryParameters: null);
    IdResponse idResponse = IdResponse.fromJson(res.response);
    return idResponse.id;
  }
}

File: apps/client/src/app/core/services/article.service.ts
import { environment } from '../../../environments/environment';
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ArticlesResponse } from '@shared/responses/articles.response';
import { ArticleResponse } from '@shared/responses/article.response';
import { BooleanResponse } from '@shared/responses/boolean.response';
import { IdResponse } from '@shared/responses/id.response';
import { ArticleCreate } from '@shared/models/article-create.model';
import { map } from 'rxjs/operators';
import { Article } from '@shared/models/article.model';
import { ArticleEdit } from '@shared/models/article-edit.model';

@Injectable({
  providedIn: 'root'
})
export class ArticleService {

  constructor(private http: HttpClient) { }

  getArticles(): Observable<Article[]> {
    return this.http.get<ArticlesResponse>(environment.apiUrl + 'article')
      .pipe(
        map((response: ArticlesResponse) => {
          return response.data;
        })
      );
  }

  getArticle(articleId: string): Observable<Article> {
    return this.http.get<ArticleResponse>(environment.apiUrl + 'article/' + articleId)
    .pipe(
      map((response: ArticleResponse) => {
        return response.data;
      })
    );
  }

  isUniqueTitle(title: string, articleId?: string): Observable<boolean> {
    return this.http.post<BooleanResponse>(environment.apiUrl + 'article/isUniqueTitle', {
      "title": title,
      "articleId": articleId
    })
      .pipe(
        map((response: BooleanResponse) => {
          return response.data;
        })
      );
  }

  createArticle(articleToAdd: ArticleCreate): Observable<number> {
    return this.http.post<IdResponse>(environment.apiUrl + 'article', articleToAdd)
      .pipe(
        map((response: IdResponse) => {
          return response.data;
        })
      );
  }

  editArticle(articleToEdit: ArticleEdit): Observable<number> {
    return this.http.put<IdResponse>(environment.apiUrl + 'article/' + articleToEdit.id, articleToEdit)
      .pipe(
        map((response: IdResponse) => {
          return response.data;
        })
      );
  }

  delete(articleId: string): Observable<boolean> {
    return this.http.delete<BooleanResponse>(environment.apiUrl + 'article/' + articleId)
    .pipe(
      map((response: BooleanResponse) => {
        return response.data;
      })
    );
  }

}